daml-script/dump/src/main/scala/com/daml/script/dump/Encode.scala

```diff
@@ -230,21 +243,34 @@ private[dump] object Encode {
       ) + Doc.text("] tree")
   }
 
-  private[dump] def encodeCreatedEvent(
+  private[dump] def encodeSubmitCreatedEvents(
       partyMap: Map[Party, String],
       cidMap: Map[ContractId, String],
       cidRefs: Set[ContractId],
-      createdEvent: CreatedEvent,
+      evs: Seq[CreatedEvent],
   ): Doc = {
-    val createCmd =
-      Doc.text("createCmd ") + encodeRecord(partyMap, cidMap, createdEvent.getCreateArguments)
-    val cid = ContractId(createdEvent.contractId)
-    val bind = if (cidRefs.contains(cid)) { Doc.text(cidMap(cid)) + Doc.text(" <- ") }
-    else { Doc.empty }
-    val submitters = Party.subst(createdEvent.signatories)
-    (bind + Doc.text("submitMulti ") + encodeParties(partyMap, submitters) + Doc.text(
-      " [] do"
-    ) / createCmd).hang(2)
+    val submitters = evs.flatMap(ev => evParties(Kind.Created(ev))).toSet
+    val cids = ContractId.subst(evs.map(_.contractId))
+    val referencedCids = cids.filter(cid => cidRefs.contains(cid))
+    val encodedCids = referencedCids.map(cid => encodeCid(cidMap, cid))
+    val (bind, returnStmt) = referencedCids.length match {
+      case 0 => (Doc.empty, Doc.empty)
+      case 1 if referencedCids.last == cids.last => (encodedCids.last :+ " <- ", Doc.empty)
+      case 1 => (encodedCids.last :+ " <- ", "pure " +: encodedCids.last)
+      case _ => (tuple(encodedCids) :+ " <- ", "pure " +: tuple(encodedCids))
+    }
+    val submit = "submitMulti " +: encodeParties(partyMap, submitters)
+    val actions = Doc.stack(evs.map { ev =>
+      val cid = ContractId(ev.contractId)
+      val bind = if (returnStmt.nonEmpty && cidRefs.contains(cid)) {
+        encodeCid(cidMap, cid) :+ " <- "
+      } else {
+        Doc.empty
+      }
+      bind + encodeCreatedEvent(partyMap, cidMap, ev)
+    })
```

Comment:
It looks like there is some potential to factor this into something that works for generic expressions that return applicative results. Doesn’t have to be in this PR but I think it might bmake the logic a bit clearer.